MAINT: Improve float16 and float32 printing

f2013519 opened the pull request:
This patch fixes https://github.com/numpy/numpy/issues/28679.

Summary of changes: Update the max exponent cutoff for float16 and float32 (and corresponding complex) types to avoid weirdness when float types can no longer exactly represent integers.

Before the change:
print(np.half(65504)) # 65500

After the change:
print(np.half(65504)) # 6.55e+04

For a detailed background on the motivation, please check the linked issue.

f2013519 committed:
MAINT: Improve float16 and float32 printing

f2013519 committed:
Fix Typo

f2013519 commented:
@mhvk @tylerjereddy, PTAL, thanks!

mhvk reviewed (commented):
Thanks! It looks good, though I have some minor comments. Also a suggestion: let's include a comment in both places cross-referencing each other, so that if we ever change one, we will not forget to change the other too.

tylerjereddy commented on numpy/_core/tests/test_scalarprint.py:
```diff
@@ -48,6 +48,17 @@ def check(v):
+        assert_equal(str(np.single(0.000099)), "9.9e-05")
+        assert_equal(str(np.single(-0.000100001)), "-0.000100001")
+        assert_equal(str(np.single(9999999)), "9999999.0")
+        assert_equal(str(np.single(-10000000)), "-1e+07")
```
If we're going to make substantial changes to the scientific notation transition thresholds for both arrays and scalars, do we not want them to match each other as well (i.e., arrays and "scalars" start using scientific notation at the same value thresholds)?

I modified my property-based test from gh-28696 to scan over the scientific notation transitions and look for mismatches, and it seems to find them pretty easily:

<details>
<summary> sample test on this branch </summary>

```diff
--- a/numpy/_core/tests/test_scalarprint.py
+++ b/numpy/_core/tests/test_scalarprint.py
@@ -10,6 +10,10 @@
 import numpy as np
 from numpy.testing import (
     assert_, assert_equal, assert_raises, assert_raises_regex, IS_MUSL)
+import hypothesis.strategies as st
+from hypothesis.extra import numpy as hynp
+from hypothesis import given
+
 
 class TestRealScalars:
     def test_str(self):
@@ -59,6 +63,31 @@ def test_gh_28679(self):
         assert_equal(str(np.single(9999999)), "9999999.0")
         assert_equal(str(np.single(-10000000)), "-1e+07")
 
+
+    @st.composite
+    def dtype_and_val(draw):
+        # produce integer values within the bounds
+        # of a given NumPy floating point dtype
+        dt = draw(hynp.floating_dtypes(sizes=(16, 32, 64)))
+        lower_bound = np.finfo(dt).min
+        upper_bound = np.finfo(dt).max
+        int_val = draw(st.integers(min_value=lower_bound, max_value=upper_bound))
+        return (dt, int_val)
+
+
+    @given(dtype_and_val())
+    def test_gh_28679_arr_match(self, dt_val):
+        # ensure that scalars and arrays transition to
+        # scientific notation identically for a wide range
+        # of integer->float coercions
+        dt, int_val = dt_val
+        actual = str(np.array(int_val, dtype=dt))
+        expected = str(np.array([int_val], dtype=dt))
+        if "e" in actual:
+            assert "e" in expected
+        else:
+            assert not "e" in expected
+

```

</details>

The first falsifying example it finds is `'1e+03'` for scalar and `'[1000.]'` for array. If the disruption of changing the string respresentations in i.e., downstream/docstrings/other places is genuinely acceptable here, what's the reason for this to be a desirable property of our scientific notation transitions?

f2013519 replied:
What does dt = draw(hynp.floating_dtypes(sizes=(16, 32, 64))) do?

If I replace with, dt = np.float16, or np.float32, the test passes. Can you provide a simple repro using np.float16 or np.float32?

Note that we are not changing the behavior for np.float64 for reasons outlined in the discussion so we would not expect the test to pass in this case.

f2013519 replied:
However, I do think ideally we should reduce the cutoff from 1e16 to 1e15 for switching to exp for scalar doubles:

>>> a = np.float64(2**53)
>>> print(a)
9007199254740992.0
>>> a = np.float64(2**53 + 1)
>>> print(a)
9007199254740992.0
>>> np.finfo(np.float64).precision
15

Let me know your thoughts ...

tylerjereddy reviewed (commented):
I added one stress testing-related comment where I'm not fully convinced just yet. I'm sure Marten knows more than I about the rest of the analysis and his comments look sensible to me in the other places (+1 for parametrization I think).

I'm going to go ahead and close gh-28696--I think it has served its purpose to refresh the discussion and refocus this to a solution the team seems to like better (and I think I agree, modulo my query about consistency).

f2013519 commented on numpy/_core/arrayprint.py:
```diff
@@ -990,11 +990,19 @@ def fillFormat(self, data):
         abs_non_zero = absolute(finite_vals[finite_vals != 0])
+
+        # consider data type while deciding the max cutoff for exp format
+        exp_cutoff_max = 1.e8
```
np.finfo(data.dtype).precision returns 6 for float32 (even though float32 has around 7 places), but we wanted to transition around 1.e7 for float.

Sure, I will move it inside the "if condition".

f2013519 replied:
I see other implementations (Julia) cutting off at 1.e6 as well. Let's go ahead with your suggestion ...

f2013519 commented on numpy/_core/src/multiarray/scalartypes.c.src:
```diff
@@ -1304,8 +1304,12 @@ legacy_@name@_format@kind@(npy_@name@ val){
+#define LONG_DOUBLE_TYPE 2
 /**begin repeat1
  * #name = float, double, longdouble#
+ * #dtype = 0, 1, 2#
```
Yes, agreed, will update.

f2013519 commented on numpy/_core/tests/test_arrayprint.py:
```diff
@@ -919,6 +919,21 @@ def test_floatmode(self):
         a = np.float64.fromhex('-1p-97')
         assert_equal(np.float64(np.array2string(a, floatmode='unique')), a)
 
+    def test_gh_28679(self):
```
I was following the style used in other places in the file. Will keep this as is for now ...

f2013519 replied:
As others have requested to update this to parametrize, I will work on that.

f2013519 committed:
Address review comments

mhvk reviewed (commented):
I think this looks good, so do my remaining comment only if there is a need to edit anything else.

I guess one larger question is whether there should be a what's-new fragment. Since this might affect downstream doctests, perhaps that is a good idea. What do others think? (it would be `doc/release/upcoming_changes/28703.change.rst`, I think; see `README.rst` in that directory).

ngoldbaum commented:
Do we need a new revision of the legacy array printing format?

Definitely needs a release note, also needs a ping to the mailing list probably, just to alert of possible downstream doctesting noise with a note to configure doc builds to use the legacy print format in doctests.

mhvk commented:
Yes, you are right, we definitely should allow people to get the old format back by setting `legacy` -- should really have thought of that myself!

f2013519 committed:
Add release note

f2013519 commented:
So we want to fallback to old behavior in case of:

numpy.set_printoptions(legacy="2.2")

correct?

ngoldbaum commented:
2.2 or older, yeah

tylerjereddy commented:
I confirmed that the property-based test passes for 16- and 32-bit types here now.

As an aside, Marten and I both asked for parametrized testing and I'd prefer if reviewer comments could be marked as resolved by the reviewers since that change was declined by author and then marked resolved. Not a big deal, but also not really a great reason for push back on it either, and certainly would prefer if the maintainer resolved it.

f2013519 committed:
Add array printing legacy version 2.2 for backward compatibility

f2013519 commented:
> I confirmed that the property-based test passes for 16- and 32-bit types here now.
> 
> As an aside, Marten and I both asked for parametrized testing and I'd prefer if reviewer comments could be marked as resolved by the reviewers since that change was declined by author and then marked resolved. Not a big deal, but also not really a great reason for push back on it either, and certainly would prefer if the maintainer resolved it.

I've unmarked it. Since people feel strongly I'll update the tests to use parameterized testing (maybe tomorrow). Pushing another incremental commit to add legacy support.

ngoldbaum commented on doc/release/upcoming_changes/28703.change.rst:
```diff
@@ -0,0 +1,2 @@
+* Printing of ``np.float16`` and ``np.float32`` scalars and arrays have been improved by
+  adjusting the transition to scientific notation based on the floating point precision.
```
This should mention the new legacy print mode setting too.

f2013519 replied:
Ack

f2013519 commented on numpy/_core/arrayprint.py:
```diff
@@ -990,11 +990,15 @@ def fillFormat(self, data):
         # choose exponential mode based on the non-zero finite values:
         abs_non_zero = absolute(finite_vals[finite_vals != 0])
+
+        # consider data type while deciding the max cutoff for exp format
```
Done

f2013519 committed:
Address remaining review comments

f2013519 committed:
Fix Typo

f2013519 committed:
Merge branch 'numpy:main' into main

f2013519 commented:
I believe I have resolved all outstanding review comments, PTAL.

Also while looking into legacy support I found get_printoptions() is broken for legacy mode 2.1 on latest. I've fixed that and added a sanity test for all supported legacy print modes with get_printoptions.

f2013519 commented:
Is there anything else that needs to be looked into before we merge this @ngoldbaum ?

ngoldbaum commented:
I'll wait for the other reviewers to give this a look.

Keep in mind that this won't be available to new users until NumPy 2.3, so there's no particular rush to get this merged before the first RC goes out.

I added the 2.3 milestone so this won't get dropped.

mhvk reviewed (approved):
This all looks great now, let's get it in! Thanks, @f2013519 - this is a small but nice change that will avoid confusion in the future!

f2013519 commented:
> This all looks great now, let's get it in! Thanks, @f2013519 - this is a small but nice change that will avoid confusion in the future!

Thanks to everyone for your feedback/suggestions!
